### PR TITLE
SONAR-6693 WS issue_filters/search rewrite WS

### PR DESCRIPTION
**Pull request by teryk**
* return user filters and all shared filters
* add the favorite info on each filter
* requires authentication
* return [SonarQube] when it's a system issue filter

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/issue/filter/SearchAction.java [hunk 1]
```diff
@@ -48,35 +55,61 @@ public SearchAction(IssueFilterService service, IssueFilterJsonWriter issueFilte
...
-      filters.addAll(sharedFiltersWithoutUserFilters);
-      ImmutableSortedSet<IssueFilterDto> allUniqueIssueFilters = FluentIterable.from(filters).toSortedSet(IssueFilterDtoIdComparator.INSTANCE);
-      json.name("issueFilters").beginArray();
-      for (IssueFilterDto favorite : allUniqueIssueFilters) {
-        issueFilterJsonWriter.write(json, favorite, userSession);
-      }
-      json.endArray();
+    json.name("issueFilters").beginArray();
+    for (IssueFilterDto issueFilter : issueFilters) {
+      issueFilterJsonWriter.write(json, new IssueFilterWithFavourite(issueFilter, isFavourite(issueFilter, userFavouritesByFilterId)), userSession);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/filter/SearchAction.java]: MAJOR Change this instance-reference to a static reference. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2209)

File: server/sonar-server/src/main/java/org/sonar/server/issue/filter/IssueFilterJsonWriter.java [hunk 2]
```diff
@@ -21,31 +21,37 @@
...
-import org.sonar.api.server.ServerSide;
 import org.sonar.api.utils.text.JsonWriter;
 import org.sonar.core.permission.GlobalPermissions;
 import org.sonar.db.issue.IssueFilterDto;
 import org.sonar.server.user.UserSession;
 
-@ServerSide
+import static com.google.common.base.Objects.firstNonNull;
+
 public class IssueFilterJsonWriter {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/filter/IssueFilterJsonWriter.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: sonar-db/src/main/java/org/sonar/db/issue/IssueFilterDao.java [hunk 3]
```diff
@@ -45,15 +46,23 @@ public IssueFilterDto selectById(long id) {
     }
   }
 
+  /**
+   * @deprecated since 5.2 use {@link #selectByUser(DbSession, String)}
+   */
+  @Deprecated
   public List<IssueFilterDto> selectByUser(String user) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueFilterDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/db/issue/IssueFilterDao.java [hunk 4]
```diff
@@ -72,27 +81,45 @@ public IssueFilterDto selectProvidedFilterByName(String name) {
     }
   }
 
+  /**
+   * @deprecated since 5.2 use {@link #selectSharedFilters(DbSession)}
+   */
+  @Deprecated
   public List<IssueFilterDto> selectSharedFilters() {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueFilterDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/db/issue/IssueFilterDao.java [hunk 5]
```diff
@@ -72,27 +81,45 @@ public IssueFilterDto selectProvidedFilterByName(String name) {
...
 
+  public List<IssueFilterDto> selectSharedFilters(DbSession session) {
+    return mapper(session).selectSharedFilters();
+  }
+
+  /**
+   * @deprecated since 5.2 use {@link #insert(DbSession, IssueFilterDto)}
+   */
+  @Deprecated
   public void insert(IssueFilterDto filter) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueFilterDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/db/issue/IssueFilterFavouriteDao.java [hunk 6]
```diff
@@ -51,16 +52,24 @@ public IssueFilterFavouriteDto selectById(long id) {
     }
   }
 
+  /**
+   * @deprecated since 5.2 use {@link #insert(DbSession, IssueFilterFavouriteDto)}
+   */
+  @Deprecated
   public void insert(IssueFilterFavouriteDto filter) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueFilterFavouriteDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 2 major
* INFO 4 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/test/java/org/sonar/server/issue/filter/SearchActionTest.java [hunk 7]
```diff
@@ -23,72 +23,107 @@
...
-import org.sonar.server.ws.WsTester;
+import org.sonar.server.ws.TestResponse;
+import org.sonar.server.ws.WsActionTester;
 
-import static com.google.common.collect.Lists.newArrayList;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
 import static org.sonar.test.JsonAssert.assertJson;
 
 public class SearchActionTest {
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/issue/filter/SearchActionTest.java]: add @Category(DbTests.class)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* INFO 4 info

Watch the comments in this conversation to review them.

**Pull request merged**